[app][log] extend diagnostics

M3ssman opened the pull request:
# Extend diagnostic output

Currently, with latest Release 0.8.1 I face situations with exceptions like this: 

```
solr_1  | 2022-09-05 10:16:48.580 ERROR (qtp820537534-142) [   fulltext_ocr] s.OcrHighlighter Could not highlight OCR content for document => java.lang.RuntimeException: Failed to parse the OCR markup, make sure your files are well-formed and your regions start/end on complete tags! (Source was: [unknown])
solr_1  | 	at com.github.dbmdz.solrocr.formats.OcrParser.<init>(OcrParser.java:111)
solr_1  | java.lang.RuntimeException: Failed to parse the OCR markup, make sure your files are well-formed and your regions start/end on complete tags! (Source was: [unknown])

...

solr_1  | 	at java.lang.Thread.run(Unknown Source) [?:?]
solr_1  | Caused by: com.ctc.wstx.exc.WstxUnexpectedCharException: Unexpected character 'T' (code 84) in prolog; expected '<'
solr_1  |  at [row,col {unknown-source}]: [1,1]
solr_1  | 	at com.ctc.wstx.sr.StreamScanner.throwUnexpectedChar(StreamScanner.java:666) ~[?:?]
solr_1  | 	at com.ctc.wstx.sr.BasicStreamReader.nextFromProlog(BasicStreamReader.java:2148) ~[?:?]
solr_1  | 	at com.ctc.wstx.sr.BasicStreamReader.next(BasicStreamReader.java:1180) ~[?:?]
solr_1  | 	at com.github.dbmdz.solrocr.formats.alto.AltoParser.seekToNextWord(AltoParser.java:229) ~[?:?]
solr_1  | 	at com.github.dbmdz.solrocr.formats.alto.AltoParser.readNext(AltoParser.java:49) ~[?:?]
solr_1  | 	at com.github.dbmdz.solrocr.formats.OcrParser.<init>(OcrParser.java:108) ~[?:?]
solr_1  | 	... 55 more
solr_1  | 2022-09-05 10:16:48.581 INFO  (qtp820537534-142) [   fulltext_ocr] o.a.s.c.S.Request webapp=/solr path=/select params={hl.snippets=100&hl.weightMatches=true&q=ocr_text:"lugna"&hl=true&hl.ocr.abso
```

Here I have only little knowledge which data causes trouble. 
The print issueing this alert contains 250+ pages with OCR (ALTO), so more diagnostic information is really appreciated. 
The 2nd line contains the unknown-source label, but maybe the first spot can reveal some more contextual information. 
(This is the line I've guessed, but I may be wrong. If there's a better place, feel free to adapt this PR for the better.)

jbaiter commented:
Thanks for the PR, this is definitely an improvement!
However, calling `content.toString()` is not a good idea, since the OCR content can be incredibly large, way too much to handle for the logging layer and dependants.

I propose instead to log the source pointer, if present, and otherwise only the first 256 charactgers of the failing OCR:

```java
            if (content.getPointer() != null) {
              log.error("Could not highlight OCR content for document at '{}'", content.getPointer(), e);
            } else {
              log.error(
                  "Could not highlight OCR content for document with OCR content '{}...'",
                  content.subSequence(0, 256), e);
            }
```

Would that help with our problem?

M3ssman commented:
@jbaiter Thanks for your suggestions!

jbaiter commented on src/main/java/solrocr/OcrHighlighter.java:
```diff
@@ -435,10 +435,18 @@ public OcrHighlightResult[] highlightOcrFields(
+            // This catch-all prevents OCR highlighting from failing the complete query, 
+            // instead users get an error message in their Solr log.
+            String sourcePtr = "n.a.";
+            if(content.getPointer() != null && content.getPointer().sources != null) {
+              List<String> pathStrs = content.getPointer().sources.stream()
+                .map(fileSource -> fileSource.path.toString()).collect(Collectors.toList());
+              sourcePtr = String.join(",", pathStrs);
```
Why not log the whole pointer? I think the offsets are invaluable for debugging if you're using them.

M3ssman replied:
But how to log the pointer in a meaningful way? 
The clazz itself contains no human-readable `toString`-implementation.

jbaiter replied:
Ouch, you're right, my bad! That should then be a part of this change set, basically the inverse of `SourcePointer.parse(String ptr)`:

```java
  @Override
  public String toString() {
    StringBuilder sb = new StringBuilder();
    for (FileSource source : this.sources) {
      if (sb.length() > 0) {
        sb.append("+");
      }
      sb.append(source.path);
      if (source.isAscii) {
        sb.append("{ascii}");
      }
      if (!source.regions.isEmpty()) {
        sb.append("[");
        for (Region region : source.regions) {
          if (sb.charAt(sb.length() - 1) != '[') {
            sb.append(",");
          }
          sb.append(region.start);
          sb.append(":");
          sb.append(region.end);
        }
        sb.append("]");
      }
    }
    return sb.toString();
  }
```

M3ssman commented:
@jbaiter Hopefully we can have a talk at this. 
Having serious trouble just to understand the main workflow of the Plugin.

M3ssman commented:
@jbaiter Thanks again for your implementation proposal. Copied this 1:1 into `SourcePointer` and now push this to the log.

By the way, I've introduced `SourcePointerTest` and did some minor refactorings to the `SourcePointer` to easy (my) understanding of workflow and data model. 

IIUC, if there's one single invalid file (i.e. deleted, unauthorized or whatever) in a very large row of OCR-Files, the highlighting is broken for the complete print. That's the current behavior according to the new tests. 
Is this really intended? 
It feels strange to throw off 1.000+ ocr-ized pages because no 1.001 has a problem.

jbaiter commented on src/main/java/com/github/dbmdz/solrocr/model/SourcePointer.java:
```diff
@@ -129,4 +105,63 @@ private static SourcePointer.Region parseRegion(String r) {
+}
+
+/** Explicite Mapper Implementation */
+class ToFileSource implements Function<String, SourcePointer.FileSource> {
```
This would be better as a static factory method `parse` on `SourcePointer.FileSource` that you can then reference in the `.map` in L86:

```java
List<FileSource> = Arrays.stream(sourceTokens).map(FileSource::parse).collect(Collectors.toList());
```

M3ssman replied:
This seems rather a philosophical question, but if you want it this way, it's fine.
One could even ask why no `SourcePointerFactory` since under the hood these are allocated from `List<FileSource>`  :slightly_smiling_face:

jbaiter commented on src/main/java/solrocr/OcrHighlighter.java:
```diff
@@ -439,11 +439,7 @@ public OcrHighlightResult[] highlightOcrFields(
-              nSourcePtr = String.valueOf(content.getPointer().sources.size());
-            }
-            log.error("Could not highlight ocr in '{}' sourceFiles", nSourcePtr, e);
+            log.error("Could not highlight ocr in source '{}'", content.getPointer(), e);
```
This does not handle the case where the user didn't use source pointers, but stores the content in the index. Refer to https://github.com/dbmdz/solr-ocrhighlighting/pull/279#issuecomment-1243352194 on what this could look like.

jbaiter commented:
One last thing: Could you also log the docId as specified in https://github.com/dbmdz/solr-ocrhighlighting/pull/279#issuecomment-1243352194?
That could really help debugging responses with large result sets.

jbaiter commented on src/main/java/solrocr/OcrHighlighter.java:
```diff
@@ -294,7 +294,9 @@ public OcrHighlightResult[] highlightOcrFields(
               + "constructor.  Perhaps you mean to call highlightWithoutSearcher?");
     }
-
+    log.debug(
+        "highlightOcrFiels names:{}, query:{}, docIDs:{}, maxPassagesOcr:{}, respHeader:{}",
+        ocrFieldNames, query, docIDs, maxPassagesOcr, respHeader);
```
That can probably be removed?

M3ssman commented:
Sorry for the inconvenience with pushing and pushing again. 
This is just because I directly coupled this branch with an internal CI/CD-pipeline which builds the Plugin as-is after the push in our development system.

jbaiter commented on src/main/java/solrocr/OcrHighlighter.java:
```diff
@@ -435,10 +437,22 @@ public OcrHighlightResult[] highlightOcrFields(
+            // instead users get an error message in their Solr log.
+            if (content.getPointer() != null) {
+              List<SourcePointer.FileSource> fSource = content.getPointer().sources;
+              if (fSource != null && fSource.get(docId) != null) {
```
I don't understand this? How does the internal Lucene docId map to the array of file sources parsed from the pointer? This is most certainly going to throw an `IndexOutOfBoundsException` in any real-world usage (where the docIds can be very large integers).

M3ssman replied:
I hoped the `docId` was somehow related to the files that get concatenated since it was in my case rather small (113 for a print with 200+ pages), but yes, that was just me misunderstanding.

jbaiter commented:
Fixed some typos and refactored the `toString()` hierarchy some more.
The squashed PR has now been merged: https://github.com/dbmdz/solr-ocrhighlighting/commit/520bf49efa4709faba5428113781c7bc3f401bd6

Somehow I messed up the squashing so GitHub doesn't register this PR as merged, but you're marked as the author and the individual commit messages are preserved in the squashed commit message, hope that's OK with you.
